hawc/apps/assessment/migrations/0031_assessment_released_on.py

```diff
@@ -0,0 +1,69 @@
+# Generated by Django 3.2.11 on 2022-01-21 17:56
+
+import json
+
+import pandas as pd
+from django.contrib.contenttypes.models import ContentType
+from django.db import migrations, models
+from django.utils import timezone
+from reversion.models import Version
+
+
+def set_released_on(apps, schema_editor):
+    ct_id = ContentType.objects.get(app_label="assessment", model="assessment").id
+
+    versions = Version.objects.filter(content_type=ct_id).select_related("revision")
+
+    data = []
+    for version in versions:
+        log = json.loads(version.serialized_data)
+        public = log[0].get("fields").get("public")
+        if public:
+            data.append([version.id, version.revision.date_created, version.object_id, public])
+
+    df = (
+        pd.DataFrame(data=data, columns="log_id revision_timestamp assessment_id public".split(" "))
+        .sort_values("revision_timestamp")
+        .reset_index(drop=True)
+    )
+
+    if df.shape[0] == 0:
+        return
+
+    df.assessment_id = df.assessment_id.astype(int)
+
+    df2 = df.groupby("assessment_id").nth(0)[["revision_timestamp"]].reset_index()
+
+    timestamps = pd.Series(data=df2.revision_timestamp.values, index=df2.assessment_id)
+
+    updates = []
+    Assessment = apps.get_model("assessment", "assessment")
+    # set the released_on date for all public assessments
+    for assessment in Assessment.objects.filter(public=True):
+        release_dt = timestamps.get(assessment.id)
+        if release_dt:
+            assessment.released_on = release_dt
+        else:  # if a date was not found, set released_on to now
+            assessment.released_on = timezone.now()
+        updates.append(assessment)
+
+    Assessment.objects.bulk_update(updates, ["released_on"])
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("assessment", "0030_assessment_models"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="assessment",
+            name="released_on",
+            field=models.DateTimeField(
+                help_text="The date this assessment was released to be viewed by the general public.",
+                null=True,
+            ),
+        ),
+        migrations.RunPython(set_released_on, migrations.RunPython.noop),
```

Comment:
might be good to add a reverse for this one. basically

```python
Assessment.objects.filter(released_on__notnull=True).update(public=True)
```

We dont generally need to run the reverse, but it can be nice to have if they're easy like this.

Reply:
good call, I overlooked that one